src/ui/training/index.ts

```diff
@@ -43,12 +42,12 @@ export default {
         .catch(handleXhrError)
       }
     } else {
-      xhr.newPuzzle()
-      .then(cfg => {
+      const onSuccess = (cfg: PuzzleData) => {
         this.ctrl = new TrainingCtrl(cfg)
         cachedState.ctrl = this.ctrl
-      })
-      .catch(handleXhrError)
+      }
+      const afterSync = () => loadOfflinePuzzle().then(onSuccess, puzzleLoadFailure)
+      syncPuzzles().then(afterSync, afterSync)
     }
```

Comment:
You can rewrite it like (betters style, to avoid `promise.then(succ, fail)`)
```
syncPuzzles()
.then(loadOfflinePuzzle)
.then((cfg: PuzzleData) => {
  this.ctrl = new TrainingCtrl(cfg)
  cachedState.ctrl = this.ctrl
})
.catch(puzzleLoadFailure)
```

Reply:
Sorry to be dumb, but can you walk me through the logic on this one? In the original version, we call `loadOfflinePuzzle` regardless of whether the call to `syncPuzzles` is successful. In the proposed version, doesn't a failure on `syncPuzzles` result in `loadOfflinePuzzle` never being called? Perhaps instead we could do this:
```
syncPuzzles(database)
        .then(() => loadOfflinePuzzle(database))
        .catch(() => loadOfflinePuzzle(database))
        .then((cfg: PuzzleData) => {
          this.ctrl = new TrainingCtrl(cfg, database)
          cachedState.ctrl = this.ctrl
        })
        .catch(puzzleLoadFailure)
 ```